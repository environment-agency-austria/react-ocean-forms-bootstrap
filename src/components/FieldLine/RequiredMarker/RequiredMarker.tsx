/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { useFormText } from 'react-ocean-forms';

import { IRequiredMarkerProps } from './RequiredMarker.types';

/**
 * Displays a * if the field is required - based on the
 * presence of react-ocean-forms validators.required
 */
export const RequiredMarker: React.FC<IRequiredMarkerProps> = (props) => {
  const { required, hidden } = props;

  const requiredTitle = useFormText('ojs_field_required');

  // Do nothing in plaintext mode
  if (hidden) { return null; }

  if (required) {
    return (
      <span className="field-required" title={requiredTitle}>
        {' '}
        *
      </span>
    );
  }

  return null;
};
