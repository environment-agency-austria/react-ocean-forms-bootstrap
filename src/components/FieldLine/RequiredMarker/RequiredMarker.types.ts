import { IFieldComponentMeta } from 'react-ocean-forms';

/**
 * Props for the RequiredMarker component
 */
export interface IRequiredMarkerProps {
  /**
   * Field meta
   * @see IFieldComponentMeta
   */
  meta: IFieldComponentMeta;
}
