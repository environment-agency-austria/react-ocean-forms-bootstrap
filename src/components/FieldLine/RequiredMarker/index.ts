export * from './RequiredMarker';
export * from './RequiredMarker.types';
