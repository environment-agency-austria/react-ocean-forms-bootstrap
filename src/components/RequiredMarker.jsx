/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { fieldMetaShape, validators as defaultValidators } from 'react-ocean-forms';

/**
 * Displays a * if the field is required - based on the
 * presence of react-ocean-forms validators.required
 */
function RequiredMarker({ validators, meta }) {
  // Do nothing in plaintext mode
  if (meta.plaintext) return null;

  if (Array.isArray(validators) && validators.includes(defaultValidators.required)) {
    const requiredTitle = meta.stringFormatter('ojs_field_required');
    return (
      <span className="field-required" title={requiredTitle}>
        {' '}
        *
      </span>
    );
  }

  return null;
}

RequiredMarker.defaultProps = {
  validators: undefined,
};

RequiredMarker.propTypes = {
  meta: fieldMetaShape.isRequired,
  validators: PropTypes.arrayOf(PropTypes.func),
};

export default RequiredMarker;
