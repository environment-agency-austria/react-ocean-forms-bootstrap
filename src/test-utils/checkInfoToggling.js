/**
 * Re-usable test suite to check the info visibility
 * toggling between InfoAlert and InfoAddonButton
 */
const checkInfoToggling = (wrapper) => {
  describe('info visible toggling', () => {
    const getVisibility = () => wrapper.find('InfoAlert').prop('visible');

    const toggleCases = [
      ['InfoAddonButton', 'onClick'],
      ['InfoAlert', 'onClose'],
    ];

    describe.each(toggleCases)('%s click', (element, prop) => {
      let visibility = getVisibility();

      it(`should toggle the info visibility on ${element} click`, () => {
        for (let i = 0; i < 2; i += 1) {
          wrapper.find(element).prop(prop)();
          const newVisibility = getVisibility();
          expect(newVisibility).toBe(!visibility);
          visibility = newVisibility;
        }
      });
    });
  });
};

export default checkInfoToggling;
