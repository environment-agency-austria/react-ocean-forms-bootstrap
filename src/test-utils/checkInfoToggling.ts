import { ShallowWrapper } from 'enzyme';

/**
 * Re-usable test suite to check the info visibility
 * toggling between InfoAlert and InfoAddonButton
 */
export const checkInfoToggling = (wrapper: ShallowWrapper): void => {
  describe('info visible toggling', () => {
    const getVisibility = (): boolean => wrapper.find('InfoAlert').prop('visible');

    const toggleCases = [
      ['InfoAddonButton', 'onClick'],
      ['InfoAlert', 'onClose'],
    ];

    describe.each(toggleCases)('%s click', (element: string, prop: string) => {
      let visibility = getVisibility();

      it(`should toggle the info visibility on ${element} click`, () => {
        for (let i = 0; i < 2; i += 1) {
          const callback: Function = wrapper.find(element).prop(prop);
          callback();

          const newVisibility = getVisibility();
          expect(newVisibility).toBe(!visibility);
          visibility = newVisibility;
        }
      });
    });
  });
};
