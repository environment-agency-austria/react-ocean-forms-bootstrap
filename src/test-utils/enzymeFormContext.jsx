/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import createMockFormatter from './createMockFormatter';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
export const createMockFormContext = registerCallback => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: createMockFormatter(),

  busy: false,
  disabled: false,
  plaintext: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a mock field meta
 */
export const createMockFieldMeta = () => ({
  valid: true,
  error: null,
  isValidating: false,
  touched: false,
  stringFormatter: createMockFormatter(),
  plaintext: false,
});

/**
 * Creates a mock field shape
 */
export const createMockField = () => ({
  id: 'mock-field',
  name: 'mock-field',
  value: 'mock-value',
  disabled: false,
  onChange: jest.fn(),
  onBlur: jest.fn(),
});
