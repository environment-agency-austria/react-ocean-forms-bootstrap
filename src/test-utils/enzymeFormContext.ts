/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IFieldComponentFieldProps, IFieldComponentMeta, IFieldState, IFormContext, TFormEventListener } from 'react-ocean-forms';

import { createMockFormatter } from './createMockFormatter';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
export const createMockFormContext = (registerCallback?: Function): IFormContext => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name: string, state: IFieldState): void => {
    if (registerCallback) {
      registerCallback(name, state);
    }
  }),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name: string, callback: TFormEventListener): void => {
    if (registerCallback) {
      registerCallback(name, callback);
    }
  }),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: createMockFormatter(),
  submit: jest.fn(),

  busy: false,
  disabled: false,
  plaintext: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a mock field meta
 */
export const createMockFieldMeta = (): IFieldComponentMeta => ({
  valid: true,
  error: null,
  isValidating: false,
  isRequired: false,
  touched: false,
  stringFormatter: createMockFormatter(),
  plaintext: false,
});

/**
 * Creates a mock field shape
 */
export const createMockField = (): IFieldComponentFieldProps => ({
  id: 'mock-field',
  name: 'mock-field',
  value: 'mock-value',
  disabled: false,
  onChange: jest.fn(),
  onBlur: jest.fn(),
});
