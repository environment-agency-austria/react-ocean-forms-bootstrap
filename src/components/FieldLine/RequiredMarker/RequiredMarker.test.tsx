import * as React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';

import { createMockFieldMeta } from '../../../test-utils/enzymeFormContext';
import { RequiredMarker } from './RequiredMarker';

describe('<RequiredMarker />', () => {
  const setup = (plaintext: boolean, required?: boolean): ShallowWrapper => {
    const meta = createMockFieldMeta();
    meta.plaintext = plaintext;
    if (required !== undefined) {
      meta.isRequired = required;
    }

    return shallow((
      <RequiredMarker
        meta={meta}
      />
    ));
  };

  it('should render nothing if meta.plaintext is active', () => {
    const wrapper = setup(true);
    expect(wrapper.find('.field-required').exists()).toBeFalsy();
  });

  it('should render nothing if there is no required field validator', () => {
    const wrapper = setup(false);
    expect(wrapper.find('.field-required').exists()).toBeFalsy();
  });

  it('should render the field required marker if there is a required field validator', () => {
    const wrapper = setup(false, true);
    expect(wrapper.find('.field-required').exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});
