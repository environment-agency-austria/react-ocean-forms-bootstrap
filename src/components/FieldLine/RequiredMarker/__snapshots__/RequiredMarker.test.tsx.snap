// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequiredMarker /> should render the field required marker if there is a required field validator 1`] = `
<span
  className="field-required"
  title="ojs_field_required"
>
   
  *
</span>
`;
